### PR TITLE
Advanced class tuning intro and grid search

### PR DESCRIPTION
**Pull request by topepo**
New data set. 

Slides mostly similar to previous. There is  an animation I took from the new book. 

New sections would be racing and sequential search.

### CODE REVIEW

File: slides/05-feature-engineering.qmd [hunk 1]
```diff
@@ -488,7 +488,7 @@ What happens for a 20% threshold?
 #| echo: false
 
 glm_val_pred %>% 
-  roc_curve(truth = on_goal, estimate = .pred_yes) %>%
```
Comment (topepo) [inline slides/05-feature-engineering.qmd:491]: Maybe these have been fixed already. I don't see how they passed in previous sessions.

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 2]
```diff
@@ -0,0 +1,697 @@
...
+cores <- parallelly::availableCores(logical = FALSE)
+cl <- parallel::makePSOCKcluster(cores)
+doParallel::registerDoParallel(cl)
+
+options(width = 200)
+
+ggplot2::theme_set(ggplot2::theme_bw())
+```
+
+## Hotel Data
```
Comment (topepo) [inline slides/advanced-01-tuning-hyperparameters.qmd:40]: I don't know if I'll do a recipes section before this. If so, I'll move these initial data slides there.

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 3]
```diff
@@ -0,0 +1,697 @@
...
+base_rec <- 
+  hotel_rec %>% 
+  step_dummy(all_nominal_predictors()) %>% 
+  step_zv(all_predictors())
+
+base_wflow <- workflow(base_rec, linear_reg())
+```
+
+
+## A Baseline Model
```
Comment (topepo) [inline slides/advanced-01-tuning-hyperparameters.qmd:160]: Just a warm up to review the syntax

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 4]
```diff
@@ -0,0 +1,697 @@
...
+. . .
+
+- Find good values for these parameters.
+
+. . .
+
+- Once the value(s) of the parameter(s) are determined, a model can be finalized by fitting the model to the entire training set.
+
+
+## Boosted Trees
```
Comment (topepo) [inline slides/advanced-01-tuning-hyperparameters.qmd:248]: I'll probably revise this and add some sort of figure(s) to demonstrate things.

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 5]
```diff
@@ -0,0 +1,697 @@
...
+#| fig-align: center
+#| out-width: 60%
+
+knitr::include_graphics("images/grid_points.svg")
+```
+
+
+## Sequential Search
+
+We could start with a few points and search the space:
```
Comment (topepo) [inline slides/advanced-01-tuning-hyperparameters.qmd:343]: Fancy!

File: index.qmd [hunk 6]
```diff
@@ -41,19 +41,17 @@ remotes::install_github("topepo/ongoal@hockeyR")
...
+### Advanced tidymodels
+
+-   01: [Tuning hyperparameters](slides/advanced-01-tuning-hyperparameters.html)
 
--   05: [Feature engineering](slides/05-feature-engineering.html)
--   06: [Tuning hyperparameters](slides/06-tuning-hyperparameters.html)
--   07: [Transit Case Study](slides/07-transit-case-study.qmd)
--   08: [Wrapping up](slides/08-wrapping-up.html)
 
 There's also a page for [slide annotations](slides/annotations.html); these are extra notes for selected slides.
```
Comment (topepo) [inline index.qmd:56]: I'm going to wait until I've got good drafts of the advanced slides before updating the annotations.

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 7]
```diff
@@ -0,0 +1,697 @@
...
+## Choose a parameter combination `r hexes(c("tune"))`
+
+Create your own tibble for final parameters or use one of the `tune::select_*()` functions:
+
+```{r}
+lgbm_best <- select_best(lgbm_res, metric = "mae")
+lgbm_best
+```
+
+## Checking Calibration `r hexes(c("tune", "probably"))`
```
Comment (topepo) [inline slides/advanced-01-tuning-hyperparameters.qmd:574]: ¯\_(ツ)_/¯

File: docs/index.html [hunk 8]
```diff
@@ -2,7 +2,7 @@
 <html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
 
 <meta charset="utf-8">
-<meta name="generator" content="quarto-1.0.38">
+<meta name="generator" content="quarto-1.3.433">
```
Comment (simonpcouch) [inline docs/index.html:5]: This is current prod Quarto--I just downloaded that version and, at the moment, rendering generates different `site_libs/` output for me. Is this to be expected?

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 9]
```diff
@@ -0,0 +1,697 @@
...
+options(width = 200)
+
+ggplot2::theme_set(ggplot2::theme_bw())
+```
+
+## Hotel Data
+
+We'll use a version of the [hotel data](https://www.sciencedirect.com/science/article/pii/S2352340918315191). We'll use [a version](https://modeldatatoo.tidymodels.org/dev/reference/data_hotel_rates.html) where we try to predict the cost of a room. 
+
+The data are in the modeldatatoo package. We'll sample down the data and refactor some columns:
```
Comment (simonpcouch) [inline slides/advanced-01-tuning-hyperparameters.qmd:44]: To save you a line:

```
We'll use [data on hotels](https://www.sciencedirect.com/science/article/pii/S2352340918315191) to predict the cost of a room. 

The [data](https://modeldatatoo.tidymodels.org/dev/reference/data_hotel_rates.html) are in the modeldatatoo package. We'll sample down the data and refactor some columns: 
```

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 10]
```diff
@@ -0,0 +1,697 @@
...
+#| label: hotel-split
+set.seed(4028)
+hotel_split <-
+  initial_split(hotel_rates, strata = avg_price_per_room)
+
+hotel_tr <- training(hotel_split)
+hotel_te <- testing(hotel_split)
+```
+
+
```
Comment (simonpcouch) [inline slides/advanced-01-tuning-hyperparameters.qmd:103]: Perhaps print side-by-side here?

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 11]
```diff
@@ -0,0 +1,697 @@
...
+
+## Your turn {transition="slide-in"}
+
+Let's take some time and investigate the _training data_. The outcome is `avg_price_per_room`. 
+
+Are there any characteristic of the data that are interesting?
```
Comment (simonpcouch) [inline slides/advanced-01-tuning-hyperparameters.qmd:109]:
```suggestion
Are there any characteristics of the data that are interesting?
```

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 12]
```diff
@@ -0,0 +1,697 @@
...
+:::
+
+::: fragment
+It will impact inference though.
+🤔
+:::
+
+
+
+# Is the random seed a tuning parameter?
```
Comment (simonpcouch) [inline slides/advanced-01-tuning-hyperparameters.qmd:228]: I really liked this section. :)

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 13]
```diff
@@ -0,0 +1,697 @@
...
+<br>
+
+We'll focus on the popular lightgbm implementation. 
+
+## Boosted Tree Tuning Parameters
+
+* The number of predictors randomly sampled at each split (aka `mtry` in $[1, ncol(x)]$ or $(0, 1]$).
+* The number of trees (`trees` in $[1, \infty]$, but usually up to thousands)
+* The number of samples needed to further split (`min_n` in $[1, n]$).
+* The rate that each tree adapts form previous iterations (`learn_rate` in $(0, \infty]$, usual maximum is 0.1).
```
Comment (simonpcouch) [inline slides/advanced-01-tuning-hyperparameters.qmd:269]:
```suggestion
* The rate that each tree adapts from previous iterations (`learn_rate` in $(0, \infty]$, usual maximum is 0.1).
```

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 14]
```diff
@@ -0,0 +1,697 @@
...
+* The number of iterations of boosting where _no improvement_ was shown (`stop_iter`)
```
Comment (simonpcouch) [inline slides/advanced-01-tuning-hyperparameters.qmd:270]:
```suggestion
* The number of iterations of boosting where _no improvement_ was shown before stopping (`stop_iter` in $[1, $`trees`$]$)
```

File: slides/advanced-01-tuning-hyperparameters.qmd (hunk 5 above)
Comment (simonpcouch) [reply slides/advanced-01-tuning-hyperparameters.qmd:343]: Hell yeah!

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 15]
```diff
@@ -0,0 +1,697 @@
...
+
+The main two strategies for optimization are:
+
+. . .
+
+- **Grid search** 💠 which tests a pre-defined set of candidate values
+
+- **Iterative search** 🌀 which suggests/estimates new values of candidate parameters to evaluate
+
+We'll focus on grid search for a while.
```
Comment (simonpcouch) [inline slides/advanced-01-tuning-hyperparameters.qmd:311]:
```suggestion
```

Since you discuss both strategies right after this but then have a "Grid Search" title slide, I'd say this is better left out.

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 16]
```diff
@@ -0,0 +1,697 @@
...
+
+## Your turn {transition="slide-in"}
+
+![](images/parsnip-flagger.jpg){.absolute top="0" right="0" width="150" height="150"}
+
+<br>
+
+What if we used early stopping instead of optimizing the number of trees? 
+
+We could set `trees = 1000` and use tune the `stop_iter` parameter.
```
Comment (simonpcouch) [inline slides/advanced-01-tuning-hyperparameters.qmd:691]:
```suggestion
We could set `trees = 1000` and tune the `stop_iter` parameter. 
```

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 17]
```diff
@@ -0,0 +1,697 @@
...
+Faceted on the expensiveness of preprocessing used.
+:::
+
+## Your turn {transition="slide-in"}
+
+![](images/parsnip-flagger.jpg){.absolute top="0" right="0" width="150" height="150"}
+
+<br>
+
+What if we used early stopping instead of optimizing the number of trees?
```
Comment (simonpcouch) [inline slides/advanced-01-tuning-hyperparameters.qmd:689]: This feels somewhat ambiguous to me but is probably fine with some spoken clarification. We mean(?):

1) Can you describe the procedural change in how this model is fitted?
2) Would we expect tuning to take longer? Shorter? The same amount of time?
3) Based on the graphs shown before, do we have any expectations about the performance of these candidates as compared to the previous ones?

**Review by simonpcouch — APPROVED** [review]
These look great.

Some bigger picture questions:

* Is a learning goal of the workshops that participants know which functionality comes from which packages? If not, do we want to load all packages up front rather than as-we-go? If so, is just leaving the hex shown in the top right sufficient background?
* For now, the intro slides don't(?) introduce `metric_set()`. Fine to use it here, but may be an argument for switching out `metric_set()` for `metrics()` rather than an explicit metric function like `roc_auc()` in the intro slides.

File: docs/index.html (hunk 8 above)
Comment (topepo) [reply docs/index.html:5]: I don't think that they would. How are they different?

File: slides/advanced-01-tuning-hyperparameters.qmd (hunk 12 above)
Comment (topepo) [reply slides/advanced-01-tuning-hyperparameters.qmd:228]: I'm thinking that we should have a slide about seeds in the intro slides.

File: docs/index.html (hunk 8 above)
Comment (simonpcouch) [reply docs/index.html:5]: Mostly all `docs/site_libs`.

Screen Shot 2023-07-03 at 3 54 32 PM

File: slides/advanced-01-tuning-hyperparameters.qmd (hunk 9 above)
Comment (EmilHvitfeldt) [reply slides/advanced-01-tuning-hyperparameters.qmd:44]: Agree with ⬆️  repeated "We’ll use a version" is a little hard

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 18]
```diff
@@ -0,0 +1,696 @@
...
+## Tuning parameters
+
+Some model or preprocessing parameters cannot be estimated directly from the data.
+
+. . .
+
+Some examples:
+
+- Tree depth in decision trees
+- Number of neighbors in a K-nearest neighbor model
```
Comment (EmilHvitfeldt) [inline slides/advanced-01-tuning-hyperparameters.qmd:189]: Since we mention that preprocessing parameters are tunable, we should list an example

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 19]
```diff
@@ -0,0 +1,696 @@
...
+
+We'll focus on the popular lightgbm implementation. 
+
+## Boosted Tree Tuning Parameters
+
+* The number of predictors randomly sampled at each split (aka `mtry` in $[1, ncol(x)]$ or $(0, 1]$).
+* The number of trees (`trees` in $[1, \infty]$, but usually up to thousands)
+* The number of samples needed to further split (`min_n` in $[1, n]$).
+* The rate that each tree adapts from previous iterations (`learn_rate` in $(0, \infty]$, usual maximum is 0.1).
+* The number of iterations of boosting where _no improvement_ was shown before stopping (`stop_iter` in $[1, $`trees`$]$)
```
Comment (EmilHvitfeldt) [inline slides/advanced-01-tuning-hyperparameters.qmd:270]:
```suggestion
* The number of iterations of boosting where _no improvement_ was shown before stopping (`stop_iter` in $[1, \text{trees}]$)
```

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 20]
```diff
@@ -0,0 +1,696 @@
...
+lgbm_wflow %>% 
+  extract_parameter_set_dials()
+
+# Individual functions: 
+trees()
+learn_rate()
+```
+
+::: fragment
+A parameter set can be updated (e.g. to change the ranges).
```
Comment (EmilHvitfeldt) [inline slides/advanced-01-tuning-hyperparameters.qmd:386]: note: some models have uses different range than the dials object provide. 

to see the ranges, OR the updated ranges one could use `pluck()`

```r
lgbm_wflow %>% 
  extract_parameter_set_dials() %>%
  pluck("object")
```

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 21]
```diff
@@ -0,0 +1,696 @@
...
+```{r}
+#| label: get-grid 
+set.seed(12)
+grid <- 
+  lgbm_wflow %>% 
+  extract_parameter_set_dials() %>% 
+  grid_latin_hypercube(size = 25)
+
+grid
+```
```
Comment (EmilHvitfeldt) [inline slides/advanced-01-tuning-hyperparameters.qmd:402]: I think it would be nice to show a visual of what a *space-filling design* looks like.  Like the one shown on the **The results** slide

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 22]
```diff
@@ -0,0 +1,696 @@
...
+
+```{r}
+#| label: lgb-finalize
+
+lgbm_wflow <- lgbm_wflow %>% finalize_workflow(lgbm_best)
+lgbm_wflow
+```
+
+
+## Running in parallel
```
Comment (EmilHvitfeldt) [inline slides/advanced-01-tuning-hyperparameters.qmd:608]: I don't know if this would fit on slides, but maybe an annotation.

Sometimes we have people that want to run a specific model type in parallel (like a random forest), but we generally advise that parallelism should be done at a higher level like we do here

**Review by EmilHvitfeldt — COMMENTED** [review]
I like what we have, the animal slide particularly is very neat!

**Review by EmilHvitfeldt — APPROVED** [review]
I like what we have, the animal slide particularly is very neat!

File: slides/advanced-01-tuning-hyperparameters.qmd (hunk 17 above)
Comment (topepo) [reply slides/advanced-01-tuning-hyperparameters.qmd:688]: I'll talk about a lot of this when we get to the slide, but for clarity: 

1. We keep training to a specific maximum number of trees and stop when the validation statistics are consistently worse. 
2. Should take much less time. 
3. Performance should be about the same, perhaps slightly worse (like 1%-2% worse). 

Since the learning rate dominates everything, an argument can be made that the number of trees isn't that consequential. There is probably an interaction between the learning rate and number of trees though.

File: slides/advanced-01-tuning-hyperparameters.qmd [hunk 23]
```diff
@@ -0,0 +1,696 @@
...
+
+
+## A Baseline Model
+
+We'll try to optimize the mean absolute error: 
+
+```{r}
+#| label: hotel-base-res
+#| cache: true
+reg_metrics <- metric_set(mae, rsq)
```
[the remark this answers is not in the document]
Comment (topepo) [reply slides/advanced-01-tuning-hyperparameters.qmd:167]: I'll speak to that during the presentation. It's just to have multiple metrics (for my interpretation). 

I've added a slide with the equations though and will add annotations on it.

File: slides/advanced-01-tuning-hyperparameters.qmd (hunk 20 above)
Comment (topepo) [reply slides/advanced-01-tuning-hyperparameters.qmd:386]: I'll add this to an annotation

File: slides/advanced-01-tuning-hyperparameters.qmd (hunk 22 above)
Comment (topepo) [reply slides/advanced-01-tuning-hyperparameters.qmd:608]: I will add an annotation and a link to an old blog post of mine about this.
